dom/src/DOMSourceFactory.ts

```diff
@@ -0,0 +1,24 @@
+import {DOMSource} from './DOMSource';
+import {SourceOptions} from './SourceOptions';
+import {MainDOMSource, MainDOMSourceOptions} from './MainDOMSource';
+import {BodyDOMSource} from './BodyDOMSource';
+import {DocumentDOMSource} from './DocumentDOMSource';
+
+export class DOMSourceFactory {
+  static create(type: string, options: DOMSourceOptions): DOMSource {
+    // Ordered most common on top for optimisation.
+    // We use if-statements instead of switch, because few conditionals
+    // optimise better with if-statements.
+    if (type !== 'document' && type !== 'body') {
+      return new MainDOMSource(<MainDOMSourceOptions>options);
+    }
+    if (type === 'body') {
+      return new BodyDOMSource(options);
+    }
+    if (type === `document`) {
+      return new DocumentDOMSource(options);
+    }
+  }
+}
+
+export type DOMSourceOptions = SourceOptions | MainDOMSourceOptions;
```

Comment:
```
export interface SourceOptions {
  runStreamAdapter: StreamAdapter;
  driverKey: string;
}

export interface MainDOMSourceOptions extends SourceOptions {
  namespace?: Array<string>;
  rootElement$: Stream<Element>;
  sanitation$: Stream<any>;
  isolateModule: IsolateModule;
  eventDelegators: Map<string, EventDelegator>;
}

export interface HTMLSourceOptions extends SourceOptions {
  html$: Stream<string>;
}

export type DOMSourceOptions = SourceOptions | MainDOMSourceOptions;
```

Saying that `HTMLSourceOptions` should extend `DOMSourceOptions` doesn’t make any sense.